### DISCUSSION TITLE
What is an element function?

### DISCUSSION THREAD

**Discussion by edgarogh** [typst/typst#1649] [Q&A]
On [the docs about "element functions"](https://typst.app/docs/reference/types/function/#element-functions):

> Some functions are associated with elements like [headings](https://typst.app/docs/reference/meta/heading/) or [tables](https://typst.app/docs/reference/layout/table/). When called, these create an element of their respective kind. In contrast to normal functions, they can further be used in [set rules](https://typst.app/docs/reference/styling/#set-rules), [show rules](https://typst.app/docs/reference/styling/#show-rules), and [selectors](https://typst.app/docs/reference/types/selector/).

This tells us what "element functions" _do_ (return elements + can be used in some places where other functions can't), but not what they _are_, i.e. what makes them element function intrinsically. There is nothing explaining what an "element" is (and the concept seems tightly coupled with functions anyway).

  * What makes a [`content`](https://typst.app/docs/reference/types/content/) value an "element"?
  * Are "elements" reserved for built-in functions?
  * Is there a way to define "elements" in user code?

Thanks in advance and thank you for this amazing piece of software!

**Comment by TabError** [comment]
I have exactly the same questions. I want to generate outlines for some custom "elements"(?) and am currently looking into selectors. The word 'element' is all over the place but not explained anywhere.
(I'm sure I'll find a way to get the outline I want with the existing features and without understanding elements. Nevertheless I'd like to understand)

https://typst.app/docs/reference/foundations/selector/

**Reply by TabError** [reply]
I just found this this discussion: https://github.com/typst/typst/discussions/2681
and thus this issue: https://github.com/typst/typst/issues/147

**Reply by TabError** [reply]
Which means: They are not implemented for custom use yet. And just yesterday the dev stated that work has not begun.

**Reply by TabError** [reply]
There is this blog entry from the developer too, which explains somehow what elements are supposed to be.

https://laurmaedje.github.io/posts/types-and-context/
